AP_NMEA_Output: add msg sentence $PASHR + refactor

magicrub opened the pull request:
add msg sentence $PASHR. This allows PhaseOne cameras to get orientation data to embed into it's images

bugeats reviewed (approved):
I can confirm this branch sends correct orientation data to the Phase One ixm-100 camera when using the simulator with an RS-232 adapter.

rmackay9 commented on libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:
```diff
@@ -163,7 +164,55 @@ void AP_NMEA_Output::update()
+                            (unsigned)gps_status_flag,
+                            (unsigned)ins_status_flag);
+    if (pashr_res == -1) {
+        free(gga);
```
This free looks incorrect. I guess it should be pashr or we'd get a memory leak.  There's an [issue](https://github.com/ArduPilot/ardupilot/issues/12514) and [PR](https://github.com/ArduPilot/ardupilot/pull/22184) which suggests we shouldn't be allocating and freeing all the time.

magicrub replied:
I agree with not constantly allocating, bad practice for our embedded stuff. Meanwhile, this frees the already allocated memory. If we get here then pashr_res is -1 which means it failed to allocate pashr. See example of GPRMC above this code

rmackay9 replied:
Ah, ok, I see.

rmackay9 commented:
It looks like a useful change to me!

peterbarker commented on libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:
```diff
@@ -163,7 +164,55 @@ void AP_NMEA_Output::update()
+    const float pitch_deg = wrap_180(degrees(ahrs.get_pitch()));
+    const float yaw_deg = wrap_360(degrees(ahrs.get_yaw()));
+    const float heave_m = 0; // instantaneous heave in meters
+    const float roll_deg_accuracy = 0; // stddev of roll_deg;
+    const float pitch_deg_accuracy = 0; // stddev of pitch_deg;
+    const float heading_deg_accuracy = 0; // stddev of yaw_deg;
```
Is this correct according to spec or should the field be empty or something?

magicrub replied:
I don't know, your guess is as good as mine. Safer to put it in than not?

peterbarker reviewed (commented):
I wonder if we need a bitmask for sending this stuff....

magicrub commented on libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:
```diff
@@ -163,7 +164,55 @@ void AP_NMEA_Output::update()
+                            "$PASHR,%s,%.2f,T,%c%.2f,%c%.2f,%c%.2f,%.3f,%.3f,%.3f,%u,%u",
+                            tstring,
+                            yaw_deg, // This is a TRUE NORTH value
+                            roll_deg<0? '-':'+', fabs(roll_deg),    // always show + or - symbol
```
I know, right?

magicrub commented on libraries/AP_NMEA_Output/AP_NMEA_Output.cpp:
```diff
@@ -163,7 +164,55 @@ void AP_NMEA_Output::update()
+    // 1 = All SPAN Post-Alignment INS Status
+    const bool ins_status_flag = ahrs.initialised() &&
+                                    ahrs.healthy() &&
+                                    (!ahrs.have_inertial_nav() || AP::ins().accel_calibrated_ok_all());
```
Whatever satisfies this....
`// 0 = All SPAN Pre-Alignment INS Status`

`// 1 = All SPAN Post-Alignment INS Status`

magicrub commented:
> I wonder if we need a bitmask for sending this stuff....

you know.. I had that all written up where it had a param enable bitmask and its own first-class library spot in AP_Vehicle... but trashed it to make sure this actually has any chance in hell of merging in any reasonable time. It also caused code bloat.

rmackay9 commented:
I also support adding a bitmask to allow choosing what data is output in the NMEA stream.  Boat users have requested that depth information is also output... obviously that's out of scope for this PR but just saying that if that param was added we might use it on Rover/Boats.

magicrub commented:
Ok, so I went a little crazy on this. As requested, I've added a param for a bitmask but then it turned into more. Here's what this branch is now:
- single (first) commit adds PASHR msg (adds 456 bytes on CubeOrange, nothing on fmuv2)
- moved AP_NMEA_Output library to a first-class library to live in AP_Vehicle instead of AP_AHRS
- refactored the whole thing to utilize the string parsing and CheckSum generation code in AP_Common/NMEA.h
- Added Param: NMEA_RATE_MS to adjust the speed. Default is 100 (10Hz).
- Added Param: NMEA_MSG_EN bitmask to enable each of the three msgs types we support
- Added support for AP_Peirph (requires serial/GCS and GPS)


Adding PASHR increased codesize by 456 on CubeOrange, no change to fmuv2
Adding everything else, including the new PASHR msg, added a total of CubeOrange:360 bytes, fmuv2:0, CubeOrange-periph-heavy:1472

magicrub commented:
@rmackay9 @WickedShell  So, we can bring in that single PASHR commit and make a new PR for the other stuff, or just keep them all together here. Your call(s).

peterbarker commented on libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:
```diff
@@ -0,0 +1,22 @@
+// Needs SerialManager + (AHRS or GPS)
+
+#ifdef HAL_BUILD_AP_PERIPH
+    #define AP_NMEA_OUTPUT_LIBRARY_REQUIREMENTS (defined(HAL_GCS_ENABLED) && HAL_GCS_ENABLED && defined(HAL_PERIPH_ENABLE_GPS))
```
`HAL_GCS_ENABLED` should always be defined?

magicrub replied:
fixed

peterbarker commented on libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:
```diff
@@ -0,0 +1,22 @@
+    #define AP_NMEA_OUTPUT_LIBRARY_REQUIREMENTS (defined(HAL_GCS_ENABLED) && HAL_GCS_ENABLED && defined(HAL_PERIPH_ENABLE_GPS))
+
+    #ifndef HAL_NMEA_OUTPUT_ENABLED
+    #define HAL_NMEA_OUTPUT_ENABLED (!HAL_MINIMIZE_FEATURES && AP_NMEA_OUTPUT_LIBRARY_REQUIREMENTS)
```
Would be nice to put the HAL_MINIMIZE_FEATURES into `minimal.inc`

magicrub replied:
in another PR...

peterbarker commented on libraries/AP_NMEA_Output/AP_NMEA_Output_config.h:
```diff
@@ -0,0 +1,22 @@
+// Needs SerialManager + (AHRS or GPS)
+
+#ifdef HAL_BUILD_AP_PERIPH
+    #define AP_NMEA_OUTPUT_LIBRARY_REQUIREMENTS (defined(HAL_GCS_ENABLED) && HAL_GCS_ENABLED && defined(HAL_PERIPH_ENABLE_GPS))
```
Could we do all of this down in AP_Periph instead?

magicrub replied:
yes, but why?

magicrub replied:
To move it to periph, the only way would be in AP_Peirph.h, and anyone else touching NMEA_Output, to include AP_NMEA_Output_config.h, then check if NMEA is enabled before including the normal AP_NMEA_Output.h.

Isn't the purpose of the _config to do this kind of stuff?

magicrub replied:
@peterbarker And here's what that looks like in Periph.h
```
#include <AP_NMEA_Output/AP_NMEA_Output_config.h>
#if HAL_NMEA_OUTPUT_ENABLED
    #if HAL_GCS_ENABLED && defined(HAL_PERIPH_ENABLE_GPS)
        #include <AP_NMEA_Output/AP_NMEA_Output.h>
    #else
        // Needs SerialManager + (AHRS or GPS)
        #error "AP_NMEA_Output requires Serial/GCS and either AHRS or GPS. Needs HAL_GCS_ENABLED and HAL_PERIPH_ENABLE_GPS"
    #endif
#endif // HAL_NMEA_OUTPUT_ENABLED
```

magicrub replied:
ok, I gave in. I moved it to periph

magicrub commented:
One major thing I just noticed on this is it now enables ```ALLOW_DOUBLE_MATH_FUNCTIONS``` on all builds. That was there before too but now it hits all the periphs and makes them 64bit maths even if the feature is disabled

magicrub commented:
looks like the allow_double_maths is probably fine

magicrub commented:
I've tested this on valgrind with a TCP loopback.
- GPS1: Serial in via sim uBlox (truth data)
- GPS2: Type = 5, NMEA
- Serial5_protocol = 20 (NMEA Out). Loopback pointing to my GPS2 IP port

GPS1 and GPS2 tracked well. It was nice to see GPS2 lag (stair step changes) as I slowed down (NMEA_RATE_MS = 1000) and have it match almost perfectly as I sped up the rate (NMEA_RATE_MS = 50).

magicrub commented:
@peterbarker @tridge  OK to merge this?

tridge commented:
@magicrub I'm happy with it, need approval from @peterbarker
